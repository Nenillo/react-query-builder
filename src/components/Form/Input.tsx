import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';

const StyledInput = styled.input`
  min-width: 160px;
  padding: 0.4rem 0.6rem;
  border: 1px solid ${colors.medium};
  border-radius: 3px;
`;

export interface InputProps {
  type: 'date' | 'number' | 'text';
  value: string;
  onChange: (value: string) => void;
  className?: string;
  disabled?: boolean;
}

export const Input: React.FC<InputProps> = ({
  type,
  value,
  onChange,
  className,
  disabled = false,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value);
  };

  return (
    <StyledInput
      type={type}
      value={value}
      onChange={handleChange}
      className={className}
      disabled={disabled}
    />
  );
};
