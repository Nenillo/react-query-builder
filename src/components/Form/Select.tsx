import React from 'react';

import styled from 'styled-components';
import { colors } from '../../constants/colors';

const StyledSelect = styled.select`
  padding: 0.2rem 0.5rem;
  height: 1.6rem;
  line-height: 1.6rem;
  border: 1px solid ${colors.medium};
  border-radius: 3px;
`;

export interface SelectProps {
  values: { value: React.ReactText; label: string }[];
  selectedValue?: React.ReactText;
  emptyValue?: string;
  onChange: (value: any) => void;
  className?: string;
}

export const Select: React.FC<SelectProps> = ({
  values,
  selectedValue,
  emptyValue,
  onChange,
  className,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onChange(event.target.value);
  };

  return (
    <StyledSelect
      onChange={handleChange}
      value={selectedValue}
      className={className}
    >
      <option value="" disabled>
        {emptyValue}
      </option>
      {values.map(({ value: optionValue, label: optionLabel }) => (
        <option value={optionValue} key={optionValue}>
          {optionLabel}
        </option>
      ))}
    </StyledSelect>
  );
};
