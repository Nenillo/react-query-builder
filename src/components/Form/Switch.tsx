import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';

const Knob = styled.div`
  position: absolute;
  width: 1.3rem;
  height: 1.3rem;
  background: white;
  border-radius: 50%;
  border: 1px solid ${colors.dark};
  box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.4);
`;

const StyledSwitch = styled.div<{ switched: boolean }>`
  width: 3rem;
  height: 1.65rem;
  position: relative;
  border-radius: 1.4rem;
  border: 1px solid ${colors.dark};
  background-color: ${({ switched }) =>
    switched ? colors.primary : colors.darker};
  transition: all 0.5s;
  cursor: pointer;

  ${Knob} {
    transition: all 0.5s;
    top: 0.1rem;
    left: ${({ switched }) => (switched ? '1.3rem' : '0.1rem')};
  }
`;

export interface SwitchProps {
  switched: boolean;
  onChange?: (value: boolean) => void;
  className?: string;
}

export const Switch: React.FC<SwitchProps> = ({
  switched,
  onChange,
  className,
}) => {
  const handleClick = () => {
    if (onChange) {
      onChange(!switched);
    }
  };

  return (
    <StyledSwitch
      data-test="Switch"
      switched={switched}
      onClick={handleClick}
      className={className}
    >
      <Knob />
    </StyledSwitch>
  );
};
