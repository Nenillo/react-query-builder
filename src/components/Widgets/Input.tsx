import React, { useContext } from 'react';
import { isReactTextArray, isUndefined } from '../../utils/types';
import { BuilderContext } from '../Builder';
import { clone } from '../../utils/clone';

interface InputProps {
  type: 'date' | 'number' | 'text';
  value: React.ReactText | React.ReactText[];
  id: string;
}

export const Input: React.FC<InputProps> = ({ type, value, id }) => {
  const { data, setData, onChange, components } = useContext(BuilderContext);

  const {
    form: { Input: InputBase },
  } = components;

  const handleChange = (value: any, index?: number) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    if (!isUndefined(index)) {
      clonedData[parentIndex].value[index] = value;
    } else {
      clonedData[parentIndex].value = value;
    }

    setData(clonedData);
    onChange(clonedData);
  };

  if (isReactTextArray(value)) {
    return (
      <>
        <InputBase
          type={type}
          value={value[0]}
          onChange={(value: any) => handleChange(value, 0)}
        />
        <InputBase
          type={type}
          value={value[1]}
          onChange={(value: any) => handleChange(value, 1)}
        />
      </>
    );
  } else {
    return <InputBase type={type} value={value} onChange={handleChange} />;
  }
};
