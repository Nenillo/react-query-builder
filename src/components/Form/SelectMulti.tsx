import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';
import { Select, SelectProps } from './Select';

const Container = styled.div`
  display: grid;
  grid-auto-columns: min-content;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
`;

const OptionContainer = styled.div`
  display: grid;
  grid-auto-columns: min-content;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
  align-self: center;
`;

const Option = styled.span`
  padding: 0.3rem 0.5rem;
  color: ${colors.dark};
  font-size: 0.7rem;
  line-height: 0.7rem;
  white-space: nowrap;
  border: 1px solid ${colors.dark};
  border-radius: 3rem;
`;

const Delete = styled.span`
  cursor: pointer;
`;

export interface SelectMultiProps
  extends Pick<SelectProps, 'onChange' | 'values'> {
  onDelete: (value: string) => void;
  selectedValue: string[];
  emptyValue?: string;
}

export const SelectMulti: React.FC<SelectMultiProps> = ({
  onChange,
  onDelete,
  selectedValue,
  emptyValue,
  values,
}) => {
  return (
    <Container>
      <Select
        onChange={onChange}
        selectedValue=""
        emptyValue={emptyValue}
        values={values}
      />
      <OptionContainer>
        {selectedValue.map((value, key) => {
          const labelIndex = values.findIndex(item => item.value === value);

          return (
            <Option key={key}>
              {values[labelIndex].label}{' '}
              <Delete onClick={() => onDelete(value)} data-test="Delete">
                [x]
              </Delete>
            </Option>
          );
        })}
      </OptionContainer>
    </Container>
  );
};
