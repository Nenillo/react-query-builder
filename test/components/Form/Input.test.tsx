import { Input } from '../../../src/components/Form/Input';
import { shallow, mount } from 'enzyme';
import React from 'react';

describe('#components/Input', () => {
  it('Tests Snapshot', () => {
    expect(
      shallow(
        <Input
          disabled={false}
          type="text"
          onChange={jest.fn()}
          value={'Test'}
        />
      )
    ).toMatchSnapshot();
  });

  it('Tests user interaction', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Input disabled={false} type="text" onChange={onChange} value={'Test'} />
    );

    const input = wrapper.find('input');

    input.simulate('focus');
    input.simulate('change', { target: { value: 'test' } });
    expect(onChange).toBeCalled();
  });
});
