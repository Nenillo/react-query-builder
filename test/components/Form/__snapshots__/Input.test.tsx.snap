// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#components/Input Tests Snapshot 1`] = `
<styled.input
  disabled={false}
  onChange={[Function]}
  type="text"
  value="Test"
/>
`;
